docs/design/safety/README.md

```diff
@@ -0,0 +1,297 @@
+detectable undefined or erroneous behavior to have [fail-stop] behavior and even
+detailed diagnostics to enable better debugging. This mode will at least provide
+similar bug [detection] capabilities to [AddressSanitizer] and
+[MemorySanitizer].
+
+[fail-stop]: /docs/design/safety/terminology.md#fail-stop
+[detection]: /docs/design/safety/terminology.md#detecting
+[AddressSanitizer]: https://clang.llvm.org/docs/AddressSanitizer.html
+[MemorySanitizer]: https://clang.llvm.org/docs/MemorySanitizer.html
+
+Carbon will also have additional build modes to provide specific, narrow
+capabilities that cannot be reasonably combined into either of the above build
+modes. Each of these is expected to be an extension of either the debug or
+release build for that specific purpose. For example:
+
+-   Debug + [ThreadSanitizer]: detection of [data-races][data-race safety].
+-   Release + specialized [hardening]: some users can afford significant
+    run-time overhead in order to use additional hardening. Carbon will have
+    several specialized build modes in this space. Hardening techniques in this
+    space include [Control-Flow Integrity (CFI)][cfi] and hardware-accelerated
+    address sanitizing ([HWASAN]).
+
+[ThreadSanitizer]: https://clang.llvm.org/docs/ThreadSanitizer.html
+[hardening]: /docs/design/safety/terminology.md#hardening
+[cfi]: https://clang.llvm.org/docs/ControlFlowIntegrity.html
+[hwasan]:
+    https://clang.llvm.org/docs/HardwareAssistedAddressSanitizerDesign.html
+
+Carbon will provide a way to disable the default hardening in release builds,
+but not in a supported way. Its use is expected to be strictly for benchmarking
+purposes.
+
+## Constraints
+
+### Probabilistic techniques likely cannot stop attacks
+
+It's expected that probabilistic techniques that can be applied at the language
+level are attackable through a variety of techniques:
+
+-   The attacker might be able to attack repeatedly until it gets through.
+-   The attacker may be able to determine when the attack would be detected and
+    only run the attack when it would not be.
+-   The attacker might be able control the test condition to make detection much
+    less likely or avoid detection completely. For example, if detection is
+    based on the last 4 bits of a memory address, an attacker may be able to
+    generate memory allocations, viewing the address and only attacking when
+    there's a collision.
+
+Hardware vulnerabilities may make these attacks easier than they might otherwise
+appear. Future hardware vulnerabilities are difficult to predict.
+
+Note this statement focuses on what can be applied to the language level. Using
+a secure hash algorithm, such as SHA256, may be used to offer probabilistic
+defense in other situations. However, the overhead of a secure hash algorithm's
+calculation is significant in the context of most things that Carbon may do at
+the language level.
```

Comment:
It seems like maybe there are two things here that are worth separating:

1) If the probability of an attack is sufficiently low (eg, a one in 2^256 chance of guessing the correct hash) then that may not be practically different from a non-probabilistic protection, from the perspective of the concerns you list above. But would your point about hardware vulnerabilities also apply here too? (An information disclosure vulnerability could defeat a technique based on the attacker needing to know a secret hash.)

2) For secure hashes in particular, computing them is just too slow for us to use as part of a hardened language mechanism in a release build mode.

Reply:
Good note. I've reworked this section a bit to make the first talk about non-cryptographic probabilistic techniques and make the strong statement there.

And here, I've just flagged that cryptographic techniques typically have untenable overhead. But there are rare cases where they may be effective such as pkeys.